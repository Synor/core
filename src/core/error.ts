/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorDatabaseErrorType = 'lock_error' | 'unlock_error'
type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorDatabaseErrorType
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'EXCEPTION'

export class SynorError extends Error {
  type: SynorErrorType
  meta: any

  constructor(
    message: string,
    meta: any = {},
    type: SynorErrorType = 'EXCEPTION'
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    this.name = this.constructor.name
    this.type = type
    this.meta = meta
  }
}

type SynorDatabaseErrorMeta = {
  lockId?: string
}

export class SynorDatabaseError extends SynorError {
  type!: SynorDatabaseErrorType
  meta!: SynorDatabaseErrorMeta

  constructor(
    type: 'lock_error' | 'unlock_error',
    meta: Required<Pick<SynorDatabaseErrorMeta, 'lockId'>>
  )

  constructor(type: SynorDatabaseErrorType, meta: SynorDatabaseErrorMeta) {
    super('Database Error', meta, type)
  }
}

export class SynorMigrationError extends SynorError {
  type!: SynorMigrationErrorType
  meta!: Partial<MigrationRecord>

  constructor(type: SynorMigrationErrorType, meta: Partial<MigrationRecord>) {
    super('Migration Error', meta, type)
  }
}

export class SynorValidationError extends SynorError {
  type!: SynorValidationErrorType
  meta!: MigrationRecord

  constructor(type: SynorValidationErrorType, meta: MigrationRecord) {
    super('Validation Error', meta, type)
  }
}
