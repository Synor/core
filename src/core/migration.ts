import { getHash } from 'core/utils/get-hash'

type SynorConfig = import('../index').SynorConfig

export type MigrationType = 'DO' | 'UNDO'
export type MigrationVersion = string

export type MigrationInfo = {
  version: MigrationVersion
  type: MigrationType
  title: string
  raw: string
}

export type MigrationInfoParser = (migration: string) => MigrationInfo

export type MigrationSource = {
  version: MigrationVersion
  type: MigrationType
  title: string
  hash: string
  body: string
}

export type MigrationRecord = {
  id: number
  version: MigrationVersion
  type: MigrationType
  title: string
  hash: string
  appliedAt: Date
  appliedBy: string
  executionTime: number
}

export type MigrationRecordState = 'applied' | 'reverted'

export type ExtendedMigrationRecord = MigrationRecord & {
  state: MigrationRecordState
  revertedBy: number | null
}

function getMigrationInfoRegex({
  do: DO,
  undo: UNDO,
  seperator: SEPERATOR
}: SynorConfig['migrationInfoNotation']): RegExp {
  const version = `[0-9]+`
  const type = [DO, UNDO].join('|')
  const title = '.+'

  return new RegExp(
    `^(${version}).(${type})${SEPERATOR}(${title})(?:\\.(.+))?$`,
    'i'
  )
}

export const getMigrationInfoParser = (
  config: SynorConfig
): MigrationInfoParser => {
  const migrationInfoRegex = getMigrationInfoRegex(config.migrationInfoNotation)

  const typeMap: Record<string, MigrationType> = {
    [config.migrationInfoNotation.do.toUpperCase()]: 'DO',
    [config.migrationInfoNotation.undo.toUpperCase()]: 'UNDO'
  }

  return migration => {
    const infoMatches = migration.match(migrationInfoRegex)

    if (infoMatches === null) {
      throw new Error('PARSE_ERROR')
    }

    const [raw, version, type, title] = infoMatches

    return {
      version,
      type: typeMap[type.toUpperCase()],
      title,
      raw
    }
  }
}

export function SynorMigration(
  { version, type, title }: MigrationInfo,
  content: Buffer
): MigrationSource {
  const migration: MigrationSource = {
    version,
    type,
    title,
    get hash() {
      return getHash(this.body)
    },
    get body() {
      return content.toString('utf8')
    }
  }

  return migration
}
