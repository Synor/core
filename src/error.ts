/* eslint-disable no-dupe-class-members */

type MigrationRecord = import('./migration').MigrationRecord

type SynorMigrationErrorType = 'not_found'
type SynorValidationErrorType = 'dirty' | 'hash_mismatch'

type SynorErrorType =
  | SynorMigrationErrorType
  | SynorValidationErrorType
  | 'invalid_filename'
  | 'exception'

type SynorErrorDataType = {
  dirty: MigrationRecord
  hash_mismatch: MigrationRecord
  not_found: Partial<MigrationRecord>
  invalid_filename: { filename: string }
  exception: Error | Record<string, any>
}

export class SynorError<
  ErrorType extends SynorErrorType = 'exception'
> extends Error {
  type: ErrorType
  data: SynorErrorDataType[ErrorType]

  /**
   * Creates a general instance of SynorError
   *
   * @param message Error message
   */
  constructor(message: string)

  /**
   * Creates a special instance of SynorError with `type` and `data`
   *
   * @param message Error message
   * @param type Error type
   * @param data Error data
   */
  constructor(
    message: string,
    type: ErrorType,
    data: SynorErrorDataType[ErrorType]
  )

  constructor(
    message: string,
    type: SynorErrorType = 'exception',
    data: SynorErrorDataType[SynorErrorType] = {}
  ) {
    super(message)

    Error.captureStackTrace(this, this.constructor)

    if (data instanceof Error) {
      this.stack = `${data.stack}\n${this.stack}`
    }

    this.name = this.constructor.name
    this.type = type as ErrorType
    this.data = data as SynorErrorDataType[ErrorType]
  }
}

export function toSynorError<T extends SynorErrorType = 'exception'>(
  error: Error | SynorError<T>
): SynorError<T> {
  if (error instanceof SynorError) {
    return error
  }

  return new SynorError(
    `Exception: ${error.message}`,
    'exception',
    error
  ) as SynorError<T>
}
