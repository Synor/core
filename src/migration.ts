import { SynorError } from './error'
import { getHash } from './utils/get-hash'

type SynorConfig = import('.').SynorConfig

export type MigrationType = 'do' | 'undo'

export type MigrationInfo = {
  version: string
  type: MigrationType
  title: string
  filename: string
  extension: string
}

export type MigrationInfoParser = (migrationFilename: string) => MigrationInfo

export type MigrationSourceContent =
  | {
      /**
       * Describes the migration using some kind of [Data Definition Language](https://en.wikipedia.org/wiki/Data_definition_language)
       */
      body: string
      run?: never
    }
  | {
      body?: never
      /**
       * This should only exist if `body` is omitted.
       * The function signature depends on the `DatabaseEngine` implementation.
       */
      run: (...params: any[]) => Promise<any>
    }

/**
 * MigrationSource is where the schema migration is stored
 */
export type MigrationSource = {
  version: string
  type: MigrationType
  title: string
  hash: string
} & MigrationSourceContent

/**
 * The state of MigrationSource
 *
 * `pending` -> still not applied to the database
 */
type MigrationSourceState = 'pending'

/**
 * MigrationSource with additional meta information
 */
export type MigrationSourceInfo = MigrationSource & {
  state: MigrationSourceState
}

/**
 * MigrationRecord is a schema migration that has already run against database
 */
export type MigrationRecord = {
  id: number
  version: string
  type: MigrationType
  title: string
  hash: string
  appliedAt: Date
  appliedBy: string
  executionTime: number
  dirty: boolean
}

/**
 * The state of MigrationRecord
 *
 * `applied`  -> currently applied on the database
 *
 * `reverted` -> was in `applied` state previously and then reverted with an `undo` migration
 */
type MigrationRecordState = 'applied' | 'reverted'

/**
 * MigrationRecord with additional meta information
 */
export type MigrationRecordInfo = MigrationRecord & {
  state: MigrationRecordState
  revertedBy: number | null
}

function getMigrationInfoRegex({
  do: DO,
  undo: UNDO,
  separator: SEPARATOR,
  extension: EXTENSION
}: SynorConfig['migrationInfoNotation']): RegExp {
  const version = `[0-9]+`
  const type = [DO, UNDO].join('|')
  const title = '[\\S ]+'

  return new RegExp(
    `^(${version}).(${type})${SEPARATOR}(${title})\\.(${EXTENSION})$`
  )
}

export const getMigrationInfoParser = (
  notation: SynorConfig['migrationInfoNotation']
): MigrationInfoParser => {
  const migrationInfoRegex = getMigrationInfoRegex(notation)

  const typeMap: Record<string, MigrationType> = {
    [notation.do]: 'do',
    [notation.undo]: 'undo'
  }

  return migrationFilename => {
    const infoMatches = migrationFilename.match(migrationInfoRegex)

    if (infoMatches === null) {
      throw new SynorError(
        `Invalid Filename: ${migrationFilename}`,
        'invalid_filename',
        { filename: migrationFilename }
      )
    }

    const [filename, version, type, title, extension] = infoMatches

    return {
      version,
      type: typeMap[type],
      title,
      filename,
      extension
    }
  }
}

export function SynorMigration(
  { version, type, title }: MigrationInfo,
  sourceContent: MigrationSourceContent
): MigrationSource {
  const migration: MigrationSource = {
    version,
    type,
    title,
    ...sourceContent,
    get hash() {
      if (typeof sourceContent.body === 'undefined') {
        // Non-null assertion is required because of TypeScript quirks
        return getHash(sourceContent.run!.toString())
      }

      return getHash(sourceContent.body)
    }
  }

  return migration
}
