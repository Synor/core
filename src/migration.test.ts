import { SynorError } from './error'
import { getMigrationInfoParser, SynorMigration } from './migration'

describe('SynorMigration', () => {
  test('returns data', () => {
    expect(
      SynorMigration(
        { version: '0', type: 'do', title: 'Test', filename: '0.do.Test.sql' },
        Buffer.from('SELECT 1;')
      )
    ).toMatchSnapshot()
  })

  describe('getMigrationInfoParser', () => {
    let migrationInfoParser: ReturnType<typeof getMigrationInfoParser>

    beforeAll(() => {
      migrationInfoParser = getMigrationInfoParser({
        do: 'do',
        undo: 'undo',
        separator: '--',
        extension: 'js|sql'
      })
    })

    test('returns migrationInfoParser function', () => {
      expect(migrationInfoParser).toBeInstanceOf(Function)
    })

    test('returned function works (for valid filename)', () => {
      expect(migrationInfoParser('001.do--Test.sql')).toMatchInlineSnapshot(`
        Object {
          "filename": "001.do--Test.sql",
          "title": "Test",
          "type": "do",
          "version": "001",
        }
      `)

      expect(migrationInfoParser('001.do--Test.js')).toMatchInlineSnapshot(`
        Object {
          "filename": "001.do--Test.js",
          "title": "Test",
          "type": "do",
          "version": "001",
        }
      `)
    })

    test('returned function throws (for invalid filename)', () => {
      try {
        migrationInfoParser('001.do--Test.json')
      } catch (error) {
        expect(error).toBeInstanceOf(SynorError)
        // #28
        expect(error.data).toMatchInlineSnapshot(`
          Object {
            "filename": "001.do--Test.json",
          }
        `)
        expect(error.type).toMatchInlineSnapshot(`"invalid_filename"`)
      }
    })
  })
})
