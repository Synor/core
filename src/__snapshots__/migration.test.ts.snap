// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SynorMigration #34 returns data (with run) 1`] = `
Object {
  "hash": "f2a389a98811d61a0bb4a171dd49b43db69a2981edfd6f727d802dd795bd3685",
  "run": [Function],
  "title": "Test",
  "type": "do",
  "version": "0",
}
`;

exports[`SynorMigration returns data (with body) 1`] = `
Object {
  "body": "SELECT 1;",
  "hash": "17db4fd369edb9244b9f91d9aeed145c3d04ad8ba6e95d06247f07a63527d11a",
  "title": "Test",
  "type": "do",
  "version": "0",
}
`;
